velox/docs/develop/task.rst

```diff
@@ -0,0 +1,256 @@
+shared data partitions. Task is maintaining shared state and enabling
+coordination between the operators.
+
+Pipelines and Drivers
+---------------------
+
+Task is responsible for converting a query plan into a set of pipelines
+consisting of multiple operators stacked on top of each other.
+See :doc:`Plan Nodes and Operators <operators>` for the mapping between plan
+nodes and operators. Each plan node with multiple sources introduces a new
+pipeline or pipelines. These plan nodes include HashJoinNode, CrossJoinNode,
+MergeJoinNode, LocalMergeNode, LocalPartitionNode. Join nodes have two sources
+and introduce one new pipeline that ends at the right-side source.
+LocalMergeNode and LocalPartitionNode introduce one pipeline for each source.
+LocalPartitionNode is an exception as it may have a single source in which case
+it still introduces a new pipeline to allow for different parallelism before
+and after. Each pipeline may run multi-threaded and the number of threads used
+for each pipeline can be different. Driver refers to a single thread of
+execution of a single pipeline.
+
+Let’s look at a few examples.
+
+A query plan with a HashJoinNode is converted into two pipelines. One pipeline
+contains a HashProbe operator. Another pipeline contains a HashBuild operator.
+HashBuild operator is the last operator in the pipeline. The number of threads
+processing HashBuild pipeline can be different from the number of threads
+processing HashProbe pipeline.
+
+.. image:: images/task-join-plan.png
+    :width: 400
+    :align: center
+
+LocalPartitionNode is used to (1) repartition the data, or (2) gather data from
+multiple threads into a single thread, or (3) split the data produced by a
+single thread to allow for multi-threaded processing. A query plan with a
+LocalPartitionNode is converted into two pipelines: a producer pipeline that
+ends with LocalPartition operator and a consumer pipeline that starts with a
+LocalExchangeSourceOperator. These pipelines may run with same or different
+parallelism allowing for N:1, 1:N and N:M local exchange.
+
+.. image:: images/task-local-exchange-plan.png
+    :width: 400
+    :align: center
+
+LocalMergeNode is used to combine multiple sorted inputs while preserving
+sortedness. A query plan with a LocalMergeNode with two sources is translated
+into three pipelines: one producer pipeline per source node and one consumer
+pipeline. Producer pipelines end with a CallbackSink operator. Consumer
+pipeline starts with a LocalMerge operator. A pipeline with a LocalMerge
+operator always runs single-threaded. Consumer pipelines produce sorted output
+and typically run single-threaded as well.
+
+.. image:: images/task-local-merge-plan.png
+    :width: 400
+    :align: center
+
+To summarize, a query plan is converted into multiple pipelines and each
+pipeline is executed using multiple Drivers. A given plan node is converted
+into one or more Operators. Some plan nodes are converted into multiple
```

Comment:
`Operators` should be the small case `operators`? There are a couple of others like this as well.